errors/errors.go

```diff
@@ -15,209 +15,389 @@ package errors
+	var target *Error
+	if !errors.As(targetI, &target) {
+		return false
 	}
-	return e.err
+	return e.tag == target.tag &&
+		e.grpcCode == target.grpcCode &&
+		e.httpCode == target.httpCode
 }
 
-// Description returns the description of the error.
-func (e *Error) Description() string {
-	if e == nil {
-		return ""
+// Allow details to be mutable and added to the error in runtime
+func (e *Error) AddDetails(details ...proto.Message) *Error {
+	e.details = append(e.details, details...)
+
+	return e
+}
+
+/*** GRPC Methods ***/
+
+// GRPCStatus returns the gRPC status.Status object.
+func (e Error) GRPCStatus() *status.Status {
+	stat := status.New(e.grpcCode, e.message)
+
+	// convert details from proto.Msg -> protoiface.MsgV1
+	var convertedDetails []protoiface.MessageV1
+	for _, detail := range e.details {
+		if v1, ok := detail.(protoiface.MessageV1); ok {
+			convertedDetails = append(convertedDetails, v1)
+		} else {
+			log.Debugf("Failed to convert error details: %s", detail)
+		}
 	}
-	if e.description != "" {
-		return e.description
+
+	if len(e.details) > 0 {
+		var err error
+		stat, err = stat.WithDetails(convertedDetails...)
+		if err != nil {
+			log.Debugf("Failed to add error details: %s to status: %s", err, stat)
+		}
 	}
-	return e.err.Error()
+
+	return stat
 }
 
-// WithErrorReason used to pass reason and
-// grpcStatus code to the Error struct.
-func WithErrorReason(reason string, grpcStatusCode codes.Code) Option {
-	return func(err *Error) {
-		err.reason = reason
-		err.grpcStatusCode = grpcStatusCode
-		err.httpCode = grpccodes.HTTPStatusFromCode(grpcStatusCode)
+/*** HTTP Methods ***/
+
+// JSONErrorValue implements the errorResponseValue interface.
+func (e Error) JSONErrorValue() []byte {
+	grpcStatus := e.GRPCStatus().Proto()
+
+	// Make httpCode human readable
+
+	// If there is no http legacy code, use the http status text
+	// This will get overwritten later if there is an ErrorInfo code
+	httpStatus := e.tag
+	if httpStatus == "" {
+		httpStatus = http.StatusText(e.httpCode)
 	}
-}
 
-// WithResourceInfo used to pass ResourceInfo to the Error struct.
-func WithResourceInfo(resourceInfo *ResourceInfo) Option {
-	return func(e *Error) {
-		e.resourceInfo = resourceInfo
+	errJSON := ErrorJSON{
+		ErrorCode: httpStatus,
+		Message:   grpcStatus.GetMessage(),
 	}
-}
 
-// WithDescription used to pass a description
-// to the Error struct.
-func WithDescription(description string) Option {
-	return func(e *Error) {
-		e.description = description
+	// Handle err details
+	details := e.details
+	if len(details) > 0 {
+		errJSON.Details = make([]any, len(details))
+		for i, detail := range details {
+			// cast to interface to be able to do type switch
+			// over all possible error_details defined
+			// https://github.com/googleapis/go-genproto/blob/main/googleapis/rpc/errdetails/error_details.pb.go
+			switch typedDetail := detail.(type) {
```

Comment:
This may be hard to maintain in the future. 2 non-binding, but strong recommendations:

1. Please write a test to ensure the completness of the cases here compared with all the possible values. This way, if the upstream library adds a new case, we won't miss it.
   - We do a pretty bad job when updating deps and generally just update them without thinking. This means if the upstream library adds a case, we are almost certainly going to miss that.
2. Consider moving this `switch` to a new function, possibly even in a new file, so it's easier to test this (also for completness) and maintain it in the future.

Reply:
is my existing test not sufficient: `TestError_JSONErrorValue`. I add one, multiple, and each of these err_details

Reply:
No, that's a different test case I'm talking about.

In this `switch` you are adding a `case` for each possible type of `error_details` that is defined in https://github.com/googleapis/go-genproto/blob/main/googleapis/rpc/errdetails/error_details.pb.go (as the comment says).

The problem is that this is a 3rd-party library we have no control over. If Google decides to add a new possible type, then we will be missing a `case` for that.

We need to add a test that checks to make sure we have a `case` defined for each possible `error_details` defined in the `github.com/googleapis/go-genproto/googleapis/rpc/errdetails` package, today and in the future.

Reply:
oh I see what you are saying now. That is partly why I don't panic when you commented a while ago about panic-ing if a user adds an unknown err_detail... I still add the err_detail if it is unknown, but I can add an explicit test for ensuring the types as you suggested

Reply:
I added `TestEnsureAllErrDetailsCovered`